src/Workspaces/Core/MSBuild/MSBuild/MSBuildProjectLoader.Worker_ResolveReferences.cs

```diff
@@ -0,0 +1,291 @@
+                            var outputFilePath = await _buildManager.TryGetOutputFilePathAsync(absoluteProjectPath, _globalProperties, cancellationToken).ConfigureAwait(false);
+                            if (!string.IsNullOrEmpty(outputFilePath) &&
+                                metadataReferenceSet.Contains(outputFilePath) &&
+                                File.Exists(outputFilePath))
+                            {
+                                continue;
+                            }
+                        }
+
+                        // If metadata is preferred or there's no loader for this project, see if the output path exists.
+                        // If it does, don't create a project reference.
+                        if (_preferMetadataForReferencedProjects)
+                        {
+                            var projectRefFileInfos = await LoadProjectFileInfosAsync(
+                                absoluteProjectPath, DiagnosticReportingOptions.IgnoreAll, cancellationToken).ConfigureAwait(false);
+
+                            var done = false;
+                            foreach (var projectRefFileInfo in projectRefFileInfos)
+                            {
+                                if (!string.IsNullOrEmpty(projectRefFileInfo.OutputFilePath) && metadataReferenceSet.Contains(projectRefFileInfo.OutputFilePath))
+                                {
+                                    // We found it!
+                                    if (File.Exists(projectRefFileInfo.OutputFilePath))
+                                    {
+                                        done = true;
+                                        break;
+                                    }
+                                }
+                            }
+
+                            if (done)
+                            {
+                                continue;
+                            }
+
+                            // We didn't find the output file path in this project's metadata references, or it doesn't exist on disk.
+                            // In that case, carry on and load the project.
+                        }
+
+                        // OK, we've got to load the project.
+                        var projectRefInfos = await LoadProjectInfosFromPathAsync(absoluteProjectPath, _discoveredProjectOptions, cancellationToken).ConfigureAwait(false);
+
+                        var projectRefInfo = metadataReferenceSet.Find(projectRefInfos);
+                        if (projectRefInfo != null)
+                        {
+                            // Don't add a reference if the project already has a reference on us. Otherwise, it will cause a circularity.
+                            if (projectRefInfo.ProjectReferences.Any(pr => pr.ProjectId == id))
+                            {
+                                // If the metadata doesn't exist on disk, we'll have to remove it from the metadata references.
```

Comment:
Why is File.Exists done in this branch but not the else? Can we unify?

Reply:
The reason is because we're replacing the metadata references with a project reference in the else branch. Here, we're just going to leave the metadata references in place -- unless they don't exist on disk. If they don't exist on disk, they are going to be UnresolvedMetadataReferences, which will blow up later when we try to create a compilation.

Note: I'm marking this particular point of feedback as rejected, but I'm doing so much clean up in this file that I expect this case will be much cleaner when I'm done.